cheroot/server.py

```diff
@@ -1773,20 +1762,93 @@ def error_log(self, msg='', level=20, traceback=False):
+        return sock
+
+    def bind_unix_socket(self, bind_addr):
+        """Create (or recreate) a UNIX socket object."""
+        if IS_WINDOWS:
+            """
+            Trying to access socket.AF_UNIX under Windows
+            causes an AttributeError.
+            """
+            raise ValueError(  # or RuntimeError?
+                'AF_UNIX sockets are not supported under Windows.'
+            )
+
+        fs_permissions = 0o777  # TODO: allow changing mode
+
+        try:
+            # Make possible reusing the socket...
+            os.unlink(self.bind_addr)
+        except OSError:
+            """
+            File does not exist, which is the primary goal anyway.
+            """
+
+        sock = self.prepare_socket(
+            bind_addr=bind_addr,
+            family=socket.AF_UNIX, type=socket.SOCK_STREAM, proto=0,
+            nodelay=self.nodelay, ssl_adapter=self.ssl_adapter,
+        )
+
+        try:
+            """Linux way of pre-populating fs mode permissions."""
+            # Allow everyone access the socket...
+            os.fchmod(sock.fileno(), fs_permissions)
+            FS_PERMS_SET = True
+        except OSError:
+            FS_PERMS_SET = False
+
+        try:
+            sock = self.bind_socket(sock, bind_addr)
+        except socket.error:
+            sock.close()
+            raise
+
+        bind_addr = self.resolve_real_bind_addr(sock)
+
+        try:
+            """FreeBSD/macOS pre-populating fs mode permissions."""
+            if not FS_PERMS_SET:
+                os.lchmod(bind_addr, fs_permissions)
+                FS_PERMS_SET = True
+        except OSError:
+            pass
+
+        if not FS_PERMS_SET:
+            self.error_log(
+                'Failed to set socket fs mode permissions',
+                level=logging.WARNING,
+            )
+
+        self.bind_addr = bind_addr
+        self.socket = sock
+        return sock
+
+    @staticmethod
+    def prepare_socket(bind_addr, family, type, proto, nodelay, ssl_adapter):
+        """Create and prepare the socket object."""
+        sock = socket.socket(family, type, proto)
+        prevent_socket_inheritance(sock)
         if not IS_WINDOWS:
             # Windows has different semantics for SO_REUSEADDR,
             # so don't set it.
             # https://msdn.microsoft.com/en-us/library/ms740621(v=vs.85).aspx
-            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        if self.nodelay and not isinstance(self.bind_addr, str):
-            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        if nodelay and not isinstance(bind_addr, str):
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
 
-        if self.ssl_adapter is not None:
-            self.socket = self.ssl_adapter.bind(self.socket)
+        if ssl_adapter is not None:
+            sock = ssl_adapter.bind(sock)
```

Comment:
I'm not sure about this, but we might want to use the SSL adapter for systemd-provided sockets. If so, maybe this should be moved to `prepare_socket()`.

Reply:
That's a fair point, I prefer to think about it outside of this PR to keep changes small.